### PR TITLE
Metal Effect Support

### PR DESCRIPTION
**Pull request by TheSpydog**
An Effects interface for Metal. Much of the code is copy-pasted from mojoshader_opengl.c with a few significant changes:

- mtlCompileEffect concatenates all the MojoShader shader output into a single string, then compiles that conglomerate into a MTLLibrary. One Effect <=> One Metal Library.
- Since Metal has no concept of a "shader program" like GL does, we store pointers to the MTLLibrary, vertex/fragment MTLFunctions, and the uniform buffers associated with each shader.
- To make sure that we don't accidentally overwrite our uniform buffers before they get a chance to draw, we have to generate a new MTLBuffer and fill its contents with the constant register values whenever an effect's state gets changed. (These extra buffers are cached to prevent excessive allocation, of course.) Each shader struct has a "UBO" struct that chooses which buffer to use at any given time. This allows for a "single" buffer to update its contents throughout a frame without just totally overwriting any previous contents. It's an ugly solution but it works. I haven't found a better way to do this.

### CODE REVIEW

**Review by flibitijibibo — CHANGES_REQUESTED** [review]
Basically the only thing I see is that some `}`s need their respective comments on them per Ryan's style. Aside from that, this looks great!

File: mojoshader_metal.c [hunk 1]
```diff
@@ -0,0 +1,996 @@
...
+                memcpy(
+                    contents + (offset * 16),
+                    &regI[4 * idx],
+                    size * 16
+                );
+                break;
+
+            case MOJOSHADER_UNIFORM_BOOL:
+                // !!! FIXME: Need a test case
+                memcpy(
```
Comment (flibitijibibo) [inline mojoshader_metal.c:493]: Actually, just noticed that this array should just be a u8 array, not an ivec or anything weird like that. (Unless the UBO forced it to be an ivec4?)

**Comment by TheSpydog** [comment]
Pushed fixes for the stuff you mentioned. If everything looks good I'll send you a hg diff! :)

**Comment by flibitijibibo** [comment]
New changes are good - pull from upstream one more time and I'll make the Hg commit!

**Comment by flibitijibibo** [comment]
Pushed! https://hg.icculus.org/icculus/mojoshader/rev/3ecd29a50d9d

**Pull request closed without merging**